File: server/src/utils/stationStockSkip.js
/**
 * يطابق منطق خصم مخزون المحطة في `stationSale.service.js` لمنتجات الجالون/القارورة.
 * يُستخدم لبيع التعبئة ولتسجيل الدين.
 */

const FILLING_SKIP_STATION_STOCK_BY_NAME = new Set([
  'Water Gallon',
  'Water Bottle',
]);

const FILLING_SKIP_STATION_STOCK_BY_NAME_LOWER = new Set(
  [...FILLING_SKIP_STATION_STOCK_BY_NAME].map((n) => n.toLowerCase())
);

export function productNameSuggestsFillingSkipStock(name) {
  if (!name || typeof name !== 'string') {
    return false;
  }
  const t = name.trim();
  if (FILLING_SKIP_STATION_STOCK_BY_NAME.has(t)) {
    return true;
  }
  const lower = t.toLowerCase();
  if (FILLING_SKIP_STATION_STOCK_BY_NAME_LOWER.has(lower)) {
    return true;
  }
  if (t.includes('جالون')) {
    return true;
  }
  if (t.includes('قارورة') || t.includes('قاروره')) {
    return true;
  }
  return false;
}

/**
 * لا يُخصم مخزون المحطة لدين الجالون/القارورة (وما يطابق الاسم).
 */
export function shouldSkipStationStockForDebtProduct(product) {
  const unitType = product.unitType;
  if (unitType === 'gallon' || unitType === 'bottle') {
    return true;
  }
  const trimmedName =
    typeof product.name === 'string' ? product.name.trim() : '';
  if (trimmedName.length > 0 && productNameSuggestsFillingSkipStock(trimmedName)) {
    return true;
  }
  return false;
}

File: lib/core/firebase/station_stock_skip.dart
import 'package:amethyst/core/station_balance/station_balance_catalog.dart';

bool productNameSuggestsFillingSkipStock(String name) {
  final String t = name.trim();
  if (t == 'Water Gallon' || t == 'Water Bottle') {
    return true;
  }
  final String lower = t.toLowerCase();
  if (lower == 'water gallon' || lower == 'water bottle') {
    return true;
  }
  if (t.contains('جالون')) {
    return true;
  }
  if (t.contains('قارورة') || t.contains('قاروره')) {
    return true;
  }
  return false;
}

bool shouldSkipStationStockForDebtLine({
  required Map<String, dynamic> product,
  int? fillingLineSlot,
  bool fillingDebt = false,
}) {
  if (fillingDebt &&
      fillingLineSlot != null &&
      kStationFillingSkipStockColumnIndices.contains(fillingLineSlot)) {
    return true;
  }
  return shouldSkipStationStockForSale(
    product: product,
    fillingSale: fillingDebt,
    fillingLineSlot: fillingLineSlot,
  );
}

bool shouldSkipStationStockForSale({
  required Map<String, dynamic> product,
  required bool fillingSale,
  int? fillingLineSlot,
}) {
  final bool skipGallonBottleColumns =
      fillingSale && fillingLineSlot != null && (fillingLineSlot == 0 || fillingLineSlot == 1);
  final String trimmedName = (product['name'] as String? ?? '').trim();
  final bool nameMatchesFillingSkip = trimmedName.isNotEmpty &&
      productNameSuggestsFillingSkipStock(trimmedName);
  final String? unitType = product['unitType'] as String?;
  return skipGallonBottleColumns ||
      (fillingSale &&
          (unitType == 'gallon' || unitType == 'bottle' || nameMatchesFillingSkip));
}
